### PR TITLE
Make Import and Export Cancelable

### LINKED ISSUE #1673: Unable to cancel content import
## Summary

I was importing a content channel and wanted to cancel. I pressed cancel and it appeared as though the import canceled but then the import bar reappeared. Note: This may be an issue using an old phone.

## System information

*Please specify the Kolibri version you were using and your operating system!*

 - Version: 0.4.0beta5.dev244 - APK
 - Operating system: Android 6.0.1 (on Samsung S5)
 - Browser: n/a (Android App)

## How to reproduce

1. Install Android App
2. Import Channel ID (I was using 93a5bfcfa2a74962be843288aefcfc0e) from the Internet.
3. Press cancel.

## Screenshots

https://goo.gl/photos/wxGgdLPx8uwfsZ6fA

## Real-life consequences (anything community should be aware of, for instance how it affects your deployment)

This seems like a useful functionality. I'm assuming it's actually canceling on the backend, so perhaps not that problematic in terms of actually being able to use the app.

### LINKED ISSUE #1809: Wrong string for local export
## Summary

When exporting to local drive, the message `Importing from curation server` is displayed, instead of what should correctly say `Exporting to local drive`. Operation seems to finishes correctly, just displays the wrong message.

## System information

 - Version: [`0.4.0-446-gd1737c5`](https://github.com/learningequality/kolibri/pull/1805)
 - Operating system: Windows 7
 - Browser: Chrome

## How to reproduce

1. Import channel
1. Try to export to local drive
1. Observe the message above the progress bar

## Screenshots

win7 start running - oracle vm virtualbox_327

### PR DESCRIPTION
**Pull request by rtibbles**
## Summary

Makes import and export cancelable.

Fix task type reporting.

## Issues addressed

Fixes #1673 

Resolves #1809

### CODE REVIEW

**Comment by benjaoming** [comment]
Hmmm :/

> 14.87% of diff hit

File: kolibri/content/management/commands/exportcontent.py [hunk 1]
```diff
@@ -26,10 +26,15 @@ def handle_async(self, *args, **options):
...
             total_bytes_to_transfer = files.aggregate(Sum('file_size'))['file_size__sum']
 
+            exported_files = []
+
             with self.start_progress(total=total_bytes_to_transfer) as overall_progress_update:
 
                 for f in files:
 
+                    if self.is_cancelled():
+                        break
```
Comment (benjaoming) [inline kolibri/content/management/commands/exportcontent.py:36]: Potentially, if it discovers that it's cancelled here, there could still be files to clean up?

File: kolibri/content/management/commands/importcontent.py [hunk 2]
```diff
@@ -70,9 +70,13 @@ def _transfer(self, method, channel_id, path=None):
             files = File.objects.all()
             total_bytes_to_transfer = files.aggregate(Sum('file_size'))['file_size__sum']
 
+            downloaded_files = []
+
             with self.start_progress(total=total_bytes_to_transfer) as overall_progress_update:
 
                 for f in files:
+                    if self.is_cancelled():
+                        break
```
Comment (benjaoming) [inline kolibri/content/management/commands/importcontent.py:79]: Same as previous comment

**Comment by rtibbles** [comment]
In my defense, @benjaoming the code I was modifying had almost no tests! Have added some now.

**Comment by rtibbles** [comment]
Codecov has not accurately updated the comment - this now increases overall coverage.

**Comment by aronasorman** [comment]
Testing now.

**Comment by aronasorman** [comment]
Cancelling is working, but are you planning to fix the UI bit in this PR? It still shows 'Download Failed' momentarily before vanishing.

This is not a huge requirement for me, as I think we should just get 0.5 out as soon as possible.

**Comment by rtibbles** [comment]
Looks like we have a few other potentially blocking issues on 0.5.x cropping up, so I'll fix the UI issue here.

**Comment by rtibbles** [comment]
Looking at the frontend code, the failure message would only be showing if the status was coming back as failed. Does a cancelled task send a failed status as its final hurrah?

**Comment by aronasorman** [comment]
Testing one more time with the merged result!

File: kolibri/tasks/api.py [hunk 3]
```diff
@@ -143,68 +169,103 @@ def localdrive(self, request):
...
+                "importcontent",
+                "local",
+                channel["id"],
+                drive.datafolder,
+                update_progress=update_progress,
+                check_for_cancel=check_for_cancel)
+    except UserCancelledError:
+        connections.close_all()  # close all DB connections (FIX for #1818)
+        for channel in drive.metadata["channels"]:
+            channel_id = channel["id"]
```
Comment (ivanistheone) [inline kolibri/tasks/api.py:219]: I'd move the `connections.close_all()` right after this line to be sure. Or add another one... basicaully juyst keep this text in your clipboard and when in doubt paste it here.

File: kolibri/tasks/api.py [hunk 4]
```diff
@@ -143,68 +169,103 @@ def localdrive(self, request):
...
+                check_for_cancel=check_for_cancel)
+            call_command(
+                "importcontent",
+                "local",
+                channel["id"],
+                drive.datafolder,
+                update_progress=update_progress,
+                check_for_cancel=check_for_cancel)
+    except UserCancelledError:
+        connections.close_all()  # close all DB connections (FIX for #1818)
```
Comment (ivanistheone) [inline kolibri/tasks/api.py:217]: This one might not be necessary, but shouldn't hurt.

File: kolibri/tasks/api.py [hunk 5]
```diff
@@ -143,68 +169,103 @@ def localdrive(self, request):
...
+            update_progress=update_progress,
+            check_for_cancel=check_for_cancel)
+        try:
+            call_command(
+                "exportcontent",
+                channel.id,
+                drive.datafolder,
+                update_progress=update_progress,
+                check_for_cancel=check_for_cancel)
+        except UserCancelledError:
```
Comment (ivanistheone) [inline kolibri/tasks/api.py:247]: move the `connections.close_all()` from five lines below here.

File: kolibri/tasks/api.py (hunk 3 above)
Comment (rtibbles) [reply kolibri/tasks/api.py:219]: I think the one after the for loop should be sufficient, no? Can't see why it wouldn't get there.

File: kolibri/tasks/api.py (hunk 5 above)
Comment (rtibbles) [reply kolibri/tasks/api.py:247]: Whoops, yeah.

**Comment by ivanistheone** [comment]
Oups, left some comments, but saw this is merged now. Probably fine as is, but would be good do a quick test on windows once released to be sure.

File: kolibri/tasks/api.py (hunk 5 above)
Comment (rtibbles) [reply kolibri/tasks/api.py:247]: Oh wait - this is removing a destination for file copy, so the file is never being opened for a connection, as it's not one of the DBs in the content/databases directory, so this is not an issue.

**Comment by benjaoming** [comment]
Github should remind you of unsubmitted reviews!! :)

**Comment by rtibbles** [comment]
For context @benjaoming, this was what I suggested implementing and others agreed with, in the interests of getting something working:
```
1) For network importing, any downloaded files will be deleted, and the content db will also be deleted. It will be a nice clean, as if you never did it, cancel.
2) For importing from local disk, the content db for every channel will be deleted, so it will look to the user like nothing got imported, but actually the content from channels that completed importing from before you cancelled will remain.
3) For exporting to local disk, only the content and database for channel you were in the middle of exporting will get deleted.
```

**Comment by benjaoming** [comment]
@rtibbles my 2 comments were just about minor edge cases where there is a theoretical chance that a file doesn't get deleted, other than there, I had no comments :)

**Pull request merged**